### PR TITLE
ipn/ipnlocal: move handling of expired nodes to LocalBackend

### PR DESCRIPTION
**Pull request by andrew-d**
In order to be able to synthesize a new NetMap when a node expires, have LocalBackend start a timer when receiving a new NetMap that fires slightly after the next node expires. Additionally, move the logic that updates expired nodes into LocalBackend so it runs on every netmap (whether received from controlclient or self-triggered).

Updates #6932

Signed-off-by: Andrew Dunham <andrew@du.nham.ca>
Change-Id: I833390e16ad188983eac29eb34cc7574f555f2f3

### CODE REVIEW

File: ipn/ipnlocal/expiry.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+// was written) such that if control (or Headscale, etc.) sends a ControlTime
+// that's sufficiently far in the past, we can safely ignore it.
+var flagExpiredPeersEpoch = time.Unix(1673373066, 0)
+
+// If the offset between the current time and the time received from control is
+// larger than this, we store an offset in our mapSession to adjust future
+// clock timings.
+const minClockDelta = 1 * time.Minute
+
+type expiryManager struct {
```
Comment (bradfitz) [inline ipn/ipnlocal/expiry.go:28]: some godoc?

File: ipn/ipnlocal/expiry.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2023 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
```
Comment (bradfitz) [inline ipn/ipnlocal/expiry.go:4]: Maybe a little file comment up here? (not attached to copyright or package line)

File: ipn/ipnlocal/local.go [hunk 3]
```diff
@@ -800,7 +803,44 @@ func (b *LocalBackend) setClientStatus(st controlclient.Status) {
 		return
 	}
 
+	// Handle node expiry in the netmap
+	var nextExpiry time.Time
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:807]: // or zero value if none

File: ipn/ipnlocal/local.go [hunk 4]
```diff
@@ -800,7 +803,44 @@ func (b *LocalBackend) setClientStatus(st controlclient.Status) {
...
+		b.em.flagExpiredPeers(st.NetMap)
+
+		// Figure out when the next node in the netmap is expiring so we can
+		// start a timer to reconfigure at that point.
+		for _, peer := range st.NetMap.Peers {
+			if peer.KeyExpiry.IsZero() {
+				continue // tagged node
+			}
+			if peer.KeyExpiry.After(nextExpiry) {
+				nextExpiry = peer.KeyExpiry
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:818]: I don't follow this bit. Don't you want nextExpiry to be the soonest expiring one, not the latest?

File: ipn/ipnlocal/local.go [hunk 5]
```diff
@@ -800,7 +803,44 @@ func (b *LocalBackend) setClientStatus(st controlclient.Status) {
...
+
 	b.mu.Lock()
+
+	// Stop the existing netmap expiry timer, if any.
+	if !nextExpiry.IsZero() {
+		if b.nmExpiryTimer != nil {
+			if !b.nmExpiryTimer.Stop() {
+				// TODO: we need to coordinate with the
+				// goroutine below to ensure that the function
+				// isn't running
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:831]: do?

**Review by bradfitz — APPROVED** [review]

File: ipn/ipnlocal/local.go (hunk 4 above)
Comment (andrew-d) [reply ipn/ipnlocal/local.go]: Gooooooood catch, yes; I got this backwards 🤦

File: ipn/ipnlocal/local.go [hunk 6]
```diff
@@ -801,6 +804,43 @@ func (b *LocalBackend) setClientStatus(st controlclient.Status) {
...
+	if st.NetMap != nil {
+		b.em.flagExpiredPeers(st.NetMap)
+
+		// Figure out when the next node in the netmap is expiring so we can
+		// start a timer to reconfigure at that point.
+		for _, peer := range st.NetMap.Peers {
+			if peer.KeyExpiry.IsZero() {
+				continue // tagged node
+			}
+			if peer.KeyExpiry.Before(nextExpiry) {
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:819]: This will never be true. Nothing is before the zero value.

You need nextExpiry.IsZero() || this.

**Comment by bradfitz** [comment]
LGTM

**Pull request merged**